qa/integration-tests/src/test/java/io/zeebe/broker/it/health/DiskSpaceRecoveryITTest.java

```diff
@@ -0,0 +1,138 @@
+            .withEnv(
+                "ZEEBE_BROKER_EXPORTERS_ELASTICSEARCH_CLASSNAME",
+                "io.zeebe.exporter.ElasticsearchExporter")
+            .withEnv("ZEEBE_BROKER_EXPORTERS_ELASTICSEARCH_ARGS_URL", elasticHost)
+            .withEnv("ZEEBE_BROKER_EXPORTERS_ELASTICSEARCH_ARGS_BULK_DELAY", "1")
+            .withEnv("ZEEBE_BROKER_EXPORTERS_ELASTICSEARCH_ARGS_BULK_SIZE", "1")
+            .withEnv("ZEEBE_BROKER_DATA_SNAPSHOTPERIOD", "1m")
+            .withEnv("ZEEBE_BROKER_DATA_LOGSEGMENTSIZE", "4MB")
+            .withEnv("ZEEBE_BROKER_DATA_DISKUSAGECOMMANDWATERMARK", "0.5")
+            .withEnv("ZEEBE_BROKER_DATA_LOGINDEXDENSITY", "1")
+            .withEnv("ZEEBE_BROKER_EXPORTERS_ELASTICSEARCH_ARGS_INDEX_MESSAGE", "true");
+
+    zeebeBroker
+        .getDockerClient()
+        .createVolumeCmd()
+        .withDriver("local")
+        .withDriverOpts(Map.of("type", "tmpfs", "device", "tmpfs", "o", "size=32m"))
+        .withName(VOLUME_NAME)
+        .exec();
+    final Volume newVolume = new Volume("/usr/local/zeebe/data");
+    zeebeBroker.withCreateContainerCmdModifier(
+        cmd ->
+            cmd.withHostConfig(cmd.getHostConfig().withBinds(new Bind(VOLUME_NAME, newVolume)))
+                .withName("zeebe-test"));
+
+    return zeebeBroker;
+  }
+
+  private static ElasticsearchContainer createElastic(final Network network) {
+    final ElasticsearchContainer container =
+        new ElasticsearchContainer(
+            "docker.elastic.co/elasticsearch/elasticsearch:"
+                + RestClient.class.getPackage().getImplementationVersion());
+
+    container.withNetwork(network).withEnv("discovery.type", "single-node");
+    container.withCreateContainerCmdModifier(cmd -> cmd.withName("elastic"));
+    return container;
+  }
+
+  @AfterClass
+  public static void tearDownClass() {
+    LoggerFactory.getLogger("Test").info(zeebeBroker.getLogs());
+    zeebeBroker.stop();
+    zeebeBroker.getDockerClient().removeVolumeCmd(VOLUME_NAME).exec();
+    elastic.stop();
+  }
+
+  @Test
+  public void shouldRecoverAfterOutOfDiskSpaceWhenExporterStarts() {
+    // given
+    client =
+        ZeebeClient.newClientBuilder()
+            .brokerContactPoint(containerIPAddress + ":" + apiPort)
+            .usePlaintext()
+            .build();
+
+    // when
+    LOG.info("Wait until broker is out of disk space");
+    await()
+        .timeout(Duration.ofMinutes(3))
+        .pollInterval(1, TimeUnit.MICROSECONDS)
+        .untilAsserted(
+            () ->
+                assertThatThrownBy(this::publishMessage)
+                    .hasRootCauseMessage(
+                        "RESOURCE_EXHAUSTED: Cannot accept requests for partition 1. Broker is out of disk space"));
+
+    elastic.start();
+
+    // then
+    await()
+        .pollInterval(Duration.ofSeconds(10))
+        .timeout(Duration.ofMinutes(5))
+        .untilAsserted(() -> assertThatCode(this::publishMessage).doesNotThrowAnyException());
```

Comment:
I'm not a fan of `doesNotThrowAnyException` - can't we assert that the message was published in some way?

Reply:
Btw I thought we fixed that publish message doesnt return anything? Shouldn't it return a response now?

Reply:
Was added with https://github.com/zeebe-io/zeebe/pull/3820

Reply:
It returns a type 'PublishMessageResponse'. But it is still empty.

Reply:
Similar as how we return deployment keys, we should return message keys - helpful to correlate things with an exporter stream, for example. @Zelldon already opened an issue for it :wink: https://github.com/zeebe-io/zeebe/issues/4794

That said, you can still assert you get an non null `PublishMessageResponse` (or whatever it is) as a result, no? Wouldn't that always indicate success?

Reply:
assert(..).isNotNull() did not work.

Reply:
Hm, curious, wasn't expecting that. Well I'm fine leaving it like this, but I would propose we try to avoid it in general.